kernel/src/drivers/ps2/io.rs

```diff
@@ -6,127 +6,148 @@ static COMMAND_PORT: IOPort = unsafe { IOPort::new(0x64) };
 
 pub const WAIT_TIMEOUT: u16 = 1000;
 
-const OUTPUT_STATUS_BIT: u8 = 1 << 0;
-const INPUT_STATUS_BIT: u8 = 1 << 1;
+bitflags! {
+    pub struct StatusFlags: u8 {
+        const OUTPUT_FULL = 1 << 0;
+        const INPUT_FULL = 1 << 1;
+        const OUTPUT_PORT_2 = 1 << 5;
+    }
+}
+
+/// Represents a PS2 controller command without a return value
+#[allow(dead_code)]
+#[derive(Copy, Clone)]
+#[repr(u8)]
+pub enum ControllerCommand {
+    WriteConfig = 0x60,
+    DisablePort2 = 0xA7,
+    EnablePort2 = 0xA8,
+    DisablePort1 = 0xAD,
+    EnablePort1 = 0xAE,
+    WriteInputPort2 = 0xD4,
+}
+
+/// Represents a PS2 controller command with a return value
+#[allow(dead_code)] // Dead variants for completeness
+#[derive(Copy, Clone)]
+#[repr(u8)]
+pub enum ControllerReturnCommand {
+    ReadConfig = 0x20,
+    TestController = 0xAA,
+    TestPort1 = 0xAB,
+    TestPort2 = 0xA9,
+    IdentifyDevice = 0xF2,
+}
+
+/// Represents a PS2 device command opcode
+#[allow(dead_code)] // Dead variants for completeness
+#[derive(Copy, Clone)]
+#[repr(u8)]
+pub enum DeviceCommand {
+    EnableScanning = 0xF4,
+    DisableScanning = 0xF5,
+    SetDefaults = 0xF6,
+    SetScancode = 0xF0,
+    Reset = 0xFF,
+}
+
+/// Represents an error returned by PS/2
+#[derive(Copy, Clone, Debug)]
+pub enum Ps2Error {
+    NoData,
+}
 
 /// Sends a controller command without a return
-pub fn command(cmd: ControllerCommand) {
-    command_raw(cmd as u8);
+pub fn command(cmd: ControllerCommand) -> Result<(), Ps2Error> {
```

Comment:
TBH having all of these different functions is slightly strange. IMO they should be clearer-named too.

Reply:
They can't really be merged. `command_ret` reads, whereas `command` doesn't.

Reply:
Command can figure out when it should read though,  couldn't it?

Reply:
Yes, it could, but that's not a direction I want to take. You ideally don't want to be dealing with an `Optional` if you *know* it will or will not be present.

Reply:
Ok. IMO this solution is a bit weak though on a type system level -- you could abuse it by passing a returning command or a command with data to that function.